File: app/src/main/java/no/nav/pam/annonsemottak/receivers/amedia/AmediaDateConverter.java
package no.nav.pam.annonsemottak.receivers.amedia;

import org.apache.commons.lang3.StringUtils;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

/**
 * Converts date strings from Amedia API
 */
class AmediaDateConverter {

    private static final String AMEDIA_DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ssZ";


    static LocalDateTime convertDate(String date) {
        if (StringUtils.isBlank(date)) {
            return null;
        }

        date = date.replace("Z", "+0000");
        return LocalDateTime.parse(date, DateTimeFormatter.ofPattern(AMEDIA_DATE_PATTERN));
    }

    static String toStringUrlEncoded(LocalDateTime date) {

        String dateValue = date.atZone(ZoneId.of("Z")).format(DateTimeFormatter.ofPattern(AMEDIA_DATE_PATTERN));
        return dateValue
                .replace("+0000", "Z")
                .replace(":", "%5C:");
    }

    static LocalDateTime getInitialDate() {
        return LocalDateTime.of(1900, 1, 1, 0, 0, 0);
    }
}

File: src/main/java/no/nav/pam/annonsemottak/receivers/amedia/AmediaDateConverter.kt
package no.nav.pam.annonsemottak.receivers.amedia

import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter


/**
 * Converts date strings from Amedia API
 */
object AmediaDateConverter {

    @JvmStatic
    fun convertDate(date: String): LocalDateTime? = date
            .ifBlank { null }
            ?.replace("Z", "+0000")
            .let { LocalDateTime.parse(it, dateTimeFormatter) }

    @JvmStatic
    fun toStringUrlEncoded(date: LocalDateTime) = date.atZone(defaultZone).format(dateTimeFormatter).urlEncode()

}

private val defaultZone = ZoneId.of("Z")
private fun String.urlEncode() = this.replace("+0000", "Z").replace(":", "%5C:")
private const val amedia_date_pattern = "yyyy-MM-dd'T'HH:mm:ssZ"
private val dateTimeFormatter = DateTimeFormatter.ofPattern(amedia_date_pattern)
